### ISSUE TITLE
Sub-models terminology change suggestion

### ISSUE THREAD

**Issue by ledell** [tidymodels/stacks#44]
Hi there!  Very cool to see stacking in Tidymodels!  

I will take a deeper look at the package soon, but I wanted to share my thoughts about the terminology used -- specifically around the term _sub-models_.  Traditionally, in the stacking literature and community, those models are referred to as _base models_.  The term _sub-models_  has been used more specifically in the "Divide and Recombine" sense, namely, when you train models on a subsets/partitions of the original training set and then ensemble them together.  

I can provide more context or sources if you need it.

**Comment by simonpcouch** [comment]
Erin, it's great to hear from you! I really appreciated your talk at useR over the summer.

Trying to settle on the right language for this package has been a challenge, and your input here is much appreciated.  In `tune`, _sub-model_ is used to refer to a model identified by choosing specific combinations of the hyperparameters being tuned on from a `parsnip` model definition. (@juliasilge, let me know if that description feels incomplete/incorrect to you!) We had wondered if the extension of that definition to include the possibility of being ensembled warranted a more specific term. Especially if this language would be confusing for people more experienced with stacking, I'm up for that switch!

**Comment by juliasilge** [comment]
I definitely think that using language aligned with the existing literature and community is a good choice. 👍 

In [tune](https://tune.tidymodels.org/), the [idea of a sub-model and the "sub-model trick"](https://tune.tidymodels.org/articles/extras/optimizations.html#sub-model-speed-ups-1) are related to being able to evaluate many more models than you fit, for models like trees, knn, and regularized models. I am guessing that you might take advantage of this trick in setting up possible ensemble members, but it may be smart in the documentation here to separate the idea of a _sub-model_ (meaning the models you can evaluate from one fit) from the idea of a _base model_ (meaning a possible ensemble member). You may not need to emphasize the sub-models in main user-facing documentation much, relative to the ensemble member idea.

I do find this sentence pretty confusing right now FYI, whichever way you go:

> An ensemble member is a sub-model that has actually been selected (and possibly trained) for use in the ensemble (via nonzero stacking coefficients, usually) that is not regarded as resulting from a specific model definition, where-as a sub-model is an untrained candidate ensemble member.

I can't tell if a contrast is being set up between ensemble member and sub-model? Or if one is a certain of another?

**Comment by simonpcouch** [comment]
The context is very much appreciated, and thank you for the feedback on the current docs.

I'll work on clarifying this language in the next few days.🙂

**Comment by ledell** [comment]
Thanks for the reference to the "sub-model trick", @juliasilge.  FWIW, I don't see any parallel between what's happening with "sub-models" as described there, and what happens with base models inside stacked ensembles.  The sub-model trick is exploiting the fact that the sub-models (e.g. GBM with 100 trees, GBM with 500 trees) are the same algorithm type as the main model (e.g. GBM with 1,000 trees).  I think sub-model is a good name for those smaller GBMs that can be "extracted" from the final model.  However, with stacking, though possible, the final ensemble is not generally something that should be broken-down or taken apart since it was specifically trained through the metalearning process to be constructed in a certain way.  If you want to make predictions from the base models, that's totally fine, but they are independent models on their own merit, and not merely a _sub-model_, only extractable from another model.

You have also used the term, _candidate ensemble members_: I wanted to say that I agree that this is an appropriate way to refer to the base models when trying to highlight the fact that some of the base models might be zeroed out by the metaleraner if you use something like a Lasso GLM for the metalearning algorithm.  I think it's also ok to refer to the _candidate_ models as simply _base models_.  It's also ok to use the term, _constituent models_ interchangeably with _base models_, though that's more of an English-language description vs a technical term.

A few more comments:

> where-as a sub-model is an untrained candidate ensemble member

I consider base models (or anything that uses the word "model" vs "algorithm") to already be trained.  Can you explain why you used the word ["untrained"](https://github.com/tidymodels/stacks/search?q=untrained) here?

> Note that the fitting process is not sensitive to model definition membership. That is, while fitting an ensemble from a stack, the components are regarded as candidate ensemble members rather than as sub-models.

I am not sure I totally understand this.  Are you trying to highlight the difference between models that might be zeroed out ("candidates"), between models that were "selected" for (non-zero, or non-trivial) participation in the ensemble ("sub-models")?

Thanks for the discussion!

**Comment by simonpcouch** [comment]
Agreed—thanks for discussing here! This has been a really helpful nudge to make sure we revisit the documentation after having spent more time with implementation.

Re: the "untrained" language, this refers to the fact that candidates have only been trained on cross-folds/resamples of the training set by the time they're added with `add_candidates()`, and are trained on the full training set at the `fit_members()` step. You're spot on with your second interpretation!

Based on both of your feedback, though, I revisited/dropped most of the language that yall found confusing. Having spent more time with the implementation and written up explanatory material with applied examples, I'm not sure the distinctions I thought were important at the time of writing the README (especially the solely philosophical ones) are necessary to have an opinionated grammar about. The new README and associated docs use a language that's a bit closer to the actual API... the different forms of "base models" are consistently referred to as members, and before the members are fitted, they're referred to as candidate members.

Will leave this issue open for a bit longer to remind myself to give this another read over. Additional comments are welcome, of course.🙂

**Comment by topepo** [comment]
I think that using "candidate" instead of "sub-model" is fine but we do need to differentiate the different models with different tuning parameters. 

How does the terminology of "model configuration" sound to everyone when differentiating a GBM with 10 trees and a GBM model with 100 trees?

**Comment by ledell** [comment]
Thanks!  My final suggestion would be to reference the term, "base models", once to make the connection that they are the same and also mention the term "metalearning".  That will ensure a quick adoption/understanding by people who are already familiar with stacking and the common stacking terminology. 

[Homepage](https://stacks.tidymodels.org/) text suggestions in *italics below* : 

> To be used in the same ensemble, each of these model definitions must share the same resample. This rsample rset object, when paired with the model definitions, can be used to generate the tuning/fitting results objects for the candidate ensemble members with tune.  *The trained ensemble members are often referred to as "base models" in the stacking literature.*


> Then, the data stack can be evaluated using blend_predictions() to determine to how best to combine the outputs from each of the candidate members.  *In the stacking literature, this process is commonly called "metalearning".*

Lastly, the term "assessment set" is used (and i understand what it's used for in your stacking algo), but it might be worth providing some definition of that term, since it's not an standard term ("train", "validation", "test", "holdout").

**Comment by simonpcouch** [comment]
Much appreciated! Just added your suggestions to the README and help-docs.
